taskflow/algorithm/for_each.hpp

```diff
@@ -75,36 +75,50 @@ TF_FORCE_INLINE auto make_for_each_task(B b, E e, C c, P part = P()) {
   };
 }
 
+template<typename T, typename C>
+using is_index_func = std::is_invocable_r<void, C, T>;
+
+template<typename T, typename C>
+using is_range_func = std::is_invocable_r<void, C, T, T>;
+
 // Function: make_for_each_index_task
-template <typename B, typename E, typename S, typename C, typename P = DefaultPartitioner>
-TF_FORCE_INLINE auto make_for_each_index_task(B b, E e, S s, C c, P part = P()){
+template <typename T, typename C, typename P = DefaultPartitioner>
+TF_FORCE_INLINE auto make_for_each_index_task(T b, T e, C c, P part = P()){
+  using T_t = std::decay_t<unwrap_ref_decay_t<T>>;
+
+  static_assert(std::is_integral<T_t>::value, "Begin and end values must be an integral type.");
+  static_assert(
+        std::disjunction<is_index_func<T_t, C>, is_range_func<T_t, C>>::value,
```

Comment:
This might be easier to read as follows :-)
```suggestion
        is_index_func<T_t, C>::value ||  is_range_func<T_t, C>::value,
```